File: src/main/scheduler/spec.ts
import type { ScheduleSpec, Weekday } from './types'

const WEEKDAY_TO_CRON: Record<Weekday, number> = {
  sun: 0,
  mon: 1,
  tue: 2,
  wed: 3,
  thu: 4,
  fri: 5,
  sat: 6
}

export function previousTick(spec: ScheduleSpec, now: Date = new Date()): Date | null {
  if (spec.kind === 'hourly') {
    const { everyHours, minute } = spec
    const candidate = new Date(now)
    candidate.setSeconds(0, 0)
    candidate.setMinutes(minute)
    const h = candidate.getHours()
    candidate.setHours(h - (h % everyHours))
    if (candidate.getTime() > now.getTime()) {
      candidate.setHours(candidate.getHours() - everyHours)
    }
    return candidate
  }

  const dayNums = new Set(spec.days.map((d) => WEEKDAY_TO_CRON[d]))
  for (let i = 0; i < 8; i++) {
    const d = new Date(now)
    d.setDate(d.getDate() - i)
    d.setHours(spec.hour, spec.minute, 0, 0)
    if (d.getTime() <= now.getTime() && dayNums.has(d.getDay())) {
      return d
    }
  }
  return null
}

export function compileToCron(spec: ScheduleSpec): string {
  if (spec.kind === 'hourly') {
    if (!Number.isInteger(spec.everyHours) || spec.everyHours < 1 || spec.everyHours > 12) {
      throw new Error(`hourly.everyHours must be an integer in 1..12, got ${spec.everyHours}`)
    }
    if (!Number.isInteger(spec.minute) || spec.minute < 0 || spec.minute > 59) {
      throw new Error(`hourly.minute must be an integer in 0..59, got ${spec.minute}`)
    }
    const hourField = spec.everyHours === 1 ? '*' : `*/${spec.everyHours}`
    return `${spec.minute} ${hourField} * * *`
  }

  if (spec.days.length === 0) {
    throw new Error('daily.days must include at least one weekday')
  }
  if (!Number.isInteger(spec.hour) || spec.hour < 0 || spec.hour > 23) {
    throw new Error(`daily.hour must be an integer in 0..23, got ${spec.hour}`)
  }
  if (!Number.isInteger(spec.minute) || spec.minute < 0 || spec.minute > 59) {
    throw new Error(`daily.minute must be an integer in 0..59, got ${spec.minute}`)
  }

  const days = [...new Set(spec.days)]
    .map((d) => WEEKDAY_TO_CRON[d])
    .sort((a, b) => a - b)
    .join(',')
  return `${spec.minute} ${spec.hour} * * ${days}`
}

export const DEFAULT_HOURLY: Extract<ScheduleSpec, { kind: 'hourly' }> = {
  kind: 'hourly',
  everyHours: 1,
  minute: 0
}

export const DEFAULT_DAILY: Extract<ScheduleSpec, { kind: 'daily' }> = {
  kind: 'daily',
  days: ['mon', 'tue', 'wed', 'thu', 'fri'],
  hour: 9,
  minute: 0
}

File: internal/scheduler/spec.go
package scheduler

import (
	"encoding/json"
	"fmt"
	"sort"
	"strings"

	"github.com/ultrakorne/aos_cli/internal/scheduler/schedspec"
)

// Type aliases re-export the leaf schedspec package so existing callers
// (cmd/aos, tests) keep using `scheduler.ScheduleSpec` and `scheduler.Mon`.
type Weekday = schedspec.Weekday
type ScheduleSpec = schedspec.ScheduleSpec

const (
	Sun = schedspec.Sun
	Mon = schedspec.Mon
	Tue = schedspec.Tue
	Wed = schedspec.Wed
	Thu = schedspec.Thu
	Fri = schedspec.Fri
	Sat = schedspec.Sat
)

var weekdayToCron = schedspec.WeekdayToCron

// AgentMeta mirrors the .meta.json sidecar shape.
type AgentMeta struct {
	Schedule    *ScheduleSpec `json:"schedule,omitempty"`
	ScheduledAt string        `json:"scheduledAt,omitempty"`
	Title       string        `json:"title,omitempty"`
	Description string        `json:"description,omitempty"`
}

// ParseMeta parses a .meta.json sidecar; non-object content degrades to empty.
func ParseMeta(data []byte) AgentMeta {
	if len(data) == 0 {
		return AgentMeta{}
	}
	var raw map[string]json.RawMessage
	if err := json.Unmarshal(data, &raw); err != nil {
		return AgentMeta{}
	}
	var m AgentMeta
	_ = json.Unmarshal(data, &m)
	return m
}

func CompileToCron(s ScheduleSpec) (string, error) {
	switch s.Kind {
	case "hourly":
		if s.EveryHours < 1 || s.EveryHours > 12 {
			return "", fmt.Errorf("hourly.everyHours must be 1..12, got %d", s.EveryHours)
		}
		if s.Minute < 0 || s.Minute > 59 {
			return "", fmt.Errorf("hourly.minute must be 0..59, got %d", s.Minute)
		}
		hourField := "*"
		if s.EveryHours != 1 {
			hourField = fmt.Sprintf("*/%d", s.EveryHours)
		}
		return fmt.Sprintf("%d %s * * *", s.Minute, hourField), nil
	case "daily":
		if len(s.Days) == 0 {
			return "", fmt.Errorf("daily.days must include at least one weekday")
		}
		if s.Hour < 0 || s.Hour > 23 {
			return "", fmt.Errorf("daily.hour must be 0..23, got %d", s.Hour)
		}
		if s.Minute < 0 || s.Minute > 59 {
			return "", fmt.Errorf("daily.minute must be 0..59, got %d", s.Minute)
		}
		seen := map[int]struct{}{}
		nums := make([]int, 0, len(s.Days))
		for _, d := range s.Days {
			n, ok := weekdayToCron[d]
			if !ok {
				return "", fmt.Errorf("unknown weekday %q", d)
			}
			if _, dup := seen[n]; dup {
				continue
			}
			seen[n] = struct{}{}
			nums = append(nums, n)
		}
		sort.Ints(nums)
		parts := make([]string, len(nums))
		for i, n := range nums {
			parts[i] = fmt.Sprintf("%d", n)
		}
		return fmt.Sprintf("%d %d * * %s", s.Minute, s.Hour, strings.Join(parts, ",")), nil
	default:
		return "", fmt.Errorf("unknown schedule kind %q", s.Kind)
	}
}
